astropy/table/table.py

```diff
@@ -1482,7 +1482,11 @@ def _convert_col_for_table(self, col):
 
     def _init_from_cols(self, cols):
         """Initialize table from a list of Column or mixin objects."""
-        lengths = {len(col) for col in cols}
+        # Prefer shape over len if it is available (for data-less coordinates).
```

Comment:
This is not very pretty, but the alternative would be to allow `len(frame_without_data)`, which seems a bigger API break.

Reply:
I actually agree, but thought the API change would be too much. That said, it really is inconsistent to have `len(something)` not be the same as `something.shape[0]`.

Shall I go ahead nevertheless and change how `len(frame_without_data)` behaves? It certainly is true that there is perfectly good API to check whether a frame has data: `frame.has_data`...

Reply:
To elaborate on the "a bit confusing" comment. If you have a data-less `frame` with attributes of length 2, it seems very unintuitive that you can do `frame[0]` and `frame[1]` but not `len(frame)`. A common idiom is below, but speaks to the connection between indexing an object and the `len` function.
```
for ii in range(len(obj)):
    obj[ii]
```

Reply:
Yes, I agree. I'll make a separate PR that changes the behaviour for frames without data, and, if that gets accepted, will rebase this one.

Reply:
See #16833